### PR TITLE
TempTodo should be add before submit

### PR DESCRIPTION
**Pull request by przwojwwp**
[DEMO LINK](https://przwojwwp.github.io/react_todo-app-add-and-delete/)

### CODE REVIEW

**Review by natalia-klonowska — CHANGES_REQUESTED** [review]
make deploy and add demo link in pr description 

the problem is that you're adding temp todo to todos array and then replace it with createdTodo 
image

**Comment by przwojwwp** [comment]
nie mogę wyświetlić błędu przy usuwaniu todo

**Review by Zibi95 — CHANGES_REQUESTED** [review]
Hey. There is infinite loading spinner on every todo. I can't test functionalities. Nice that you are using custom hooks but there is so many of them that its hard to follow what comes from what.

**Comment by przwojwwp** [comment]
indeed, the names help a bit but it's hard to follow. Do you suggest that I should modify them so that everything is in components instead of hooks? At this stage of the task it will be very difficult

File: src/components/Footer/Footer.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      </nav>
+
+      <button
+        type="button"
+        className="todoapp__clear-completed"
+        data-cy="ClearCompletedButton"
+        disabled={completedTodosCount === 0}
+        onClick={onClearCompleted} // Add onClick handler
+      >
+        Clear completed
```
Comment (Zibi95) [inline src/components/Footer/Footer.tsx:67]: There is clearly the same pattern for every anchor. What about creating array with objects that keep the values needed to properly create every anchor with map.

**Review by Zibi95 — CHANGES_REQUESTED** [review]
Hey. You don't have too change anything about custom hooks. I'm just saying that its powerful tool but like with everything if it is overused it becomes bad. There is problem with logic of handling errors. Every error displays only once. When i try to add todo without title i get notified only once also for every TodoItem i get notified only once when i try to delete it.

File: src/components/Footer/Footer.tsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      <span className="todo-count" data-cy="TodosCounter">
+        {activeTodosCount} items left
+      </span>
+
+      <nav className="filter" data-cy="Filter">
+        {filters.map(({ type, label, cy, href }) => (
+          <a
+            key={type}
+            href={href}
+            className={`filter__link ${filter === type ? 'selected' : ''}`}
```
Comment (natalia-klonowska) [inline src/components/Footer/Footer.tsx:52]: use classnames library to improve readability of conditional classes

File: src/components/TodoItem/TodoItem.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  onErrorChange: (error: ErrorType | null) => void; // Add onErrorChange prop
+};
+
+export const TodoItem: React.FC<TodoItemProps> = ({
+  todo,
+  isTemp,
+  onErrorChange,
+}) => {
+  const { deleteTodo, isDeleting, error } = useDeleteTodo(); // Get error from useDeleteTodo
+  const [showLoader, setShowLoader] = useState<boolean>(false);
```
Comment (natalia-klonowska) [inline src/components/TodoItem/TodoItem.tsx:19]: you don't need to specify type because TypeScript can determine it from initial value

File: src/App.tsx [hunk 4]
```diff
@@ -1,26 +1,46 @@
...
-          React Todo App - Load Todos
-        </a>
-      </p>
+    <div className="todoapp">
+      <TodoContent onErrorChange={setError}>
+        <TodoList todos={filteredTodos} onErrorChange={setError} />
+        <Footer
+          todos={todos}
+          filter={filter}
+          onFilterChange={setFilter}
```
Comment (natalia-klonowska) [inline src/App.tsx:29]: you shouldn't pass state setters like setFilter and setError as props and if you're using Context you can get access these props directly inside your components instead of passing them

**Review by natalia-klonowska — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 5]
```diff
@@ -8,8 +8,7 @@ import { TodoList } from './components/TodoList/TodoList';
 import { useClearCompleted } from './hooks/useClearCompleted';
 
 const AppContent: React.FC = () => {
-  const { todos, filteredTodos, error, setFilter, setError, filter } =
-    useTodos();
+  const { todos, filteredTodos, error, filter } = useTodos();
```
Comment (natalia-klonowska) [inline src/App.tsx:11]: you don't use filter and filteredTodos you just pass it as props to other components

File: src/App.tsx [hunk 6]
```diff
@@ -21,16 +20,15 @@ const AppContent: React.FC = () => {
 
   return (
     <div className="todoapp">
-      <TodoContent onErrorChange={setError}>
-        <TodoList todos={filteredTodos} onErrorChange={setError} />
+      <TodoContent>
+        <TodoList todos={filteredTodos} />
         <Footer
           todos={todos}
           filter={filter}
```
Comment (natalia-klonowska) [inline src/App.tsx:27]: don't pass filteredTodos, todos and filter because you have it inside Context

File: src/App.tsx [hunk 7]
```diff
@@ -1,26 +1,44 @@
...
 
-const USER_ID = 0;
+const AppContent: React.FC = () => {
+  const { todos, filteredTodos, error, filter } = useTodos();
+  const { handleClearCompleted, error: clearCompletedError } =
+    useClearCompleted();
 
-export const App: React.FC = () => {
-  if (!USER_ID) {
+  const combinedError = error || clearCompletedError;
```
Comment (natalia-klonowska) [inline src/App.tsx:15]: you can move it inside Errors component

File: src/components/Errors/Errors.tsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      }
+    };
+  }, [error, setError]);
+
+  return (
+    <div
+      data-cy="ErrorNotification"
+      className={`notification is-danger is-light has-text-weight-normal ${
+        !error ? 'hidden' : ''
+      }`}
```
Comment (natalia-klonowska) [inline src/components/Errors/Errors.tsx:38]: use classnames everywhere you use conditional classes

File: src/components/TodoContent/TodoContent.tsx [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import * as React from 'react';
+import { ChangeEvent, FormEvent, useEffect, useState } from 'react';
+import { usePostTodos } from '../../hooks/usePostTodos';
+import { useTodos } from '../../utils/TodoContext';
+
+export const TodoContent: React.FC<{ children: React.ReactNode }> = ({
+  children,
+}) => {
+  const [input, setInput] = useState<string>('');
```
Comment (natalia-klonowska) [inline src/components/TodoContent/TodoContent.tsx:9]: don't need type cos it's the same as initial value (find and fix all instances where it's not needed)

**Review by natalia-klonowska — CHANGES_REQUESTED** [review]

File: src/App.tsx (hunk 5 above)
Comment (przwojwwp) [reply src/App.tsx:11]: if I change it code stop working

**Comment by przwojwwp** [comment]
never been on the theory about classnames and its syntax. If there are so many things to reorganize, I should probably do the task from the beginning. I look through other people's pull requests and these things are not implemented in them

**Comment by przwojwwp** [comment]
I cant make it myself. I need help with it

File: src/components/Footer/Footer.tsx [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+};
+
+export const Footer: React.FC<FooterProps> = ({
+  todos,
+  filter,
+  onClearCompleted,
+}) => {
+  const { setFilter } = useTodos();
+
+  if (todos.length === 0) {
```
Comment (BogdanMaliuta) [inline src/components/Footer/Footer.tsx:20]:
```suggestion
  if (!todos?.length) {
```

File: src/components/Footer/Footer.tsx [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    return null;
+  }
+
+  const activeTodosCount = todos.filter(todo => !todo.completed).length;
+  const completedTodosCount = todos.filter(todo => todo.completed).length;
+
+  const filters = [
+    { type: FilterType.All, label: 'All', cy: 'FilterLinkAll', href: '#/' },
```
Comment (BogdanMaliuta) [inline src/components/Footer/Footer.tsx:28]: split it

**Review by BogdanMaliuta — CHANGES_REQUESTED** [review]
Hello, fix what Natalia described, use your hook useTodos in components where you needs todos, if it`s stops working, just fix your hook, always remove comments from code, take a look on your types, remove it where you have initial value, and check your code style, you can improve it in some places

File: src/components/Errors/Errors.tsx [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        setClearCompletedError(null);
+        timerRef.current = null;
+      }, 3000);
+    }
+
+    return () => {
+      if (timerRef.current) {
+        clearTimeout(timerRef.current);
+        timerRef.current = null;
+      }
```
Comment (Zibi95) [inline src/components/Errors/Errors.tsx:29]: You don't need to have ref to keep timer id. You can simply define variable using `let` inside body of `effect`

File: src/components/Footer/Footer.tsx [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+export const Footer: React.FC = () => {
+  const { todos, filter, setFilter } = useTodos();
+  const { handleClearCompleted } = useClearCompleted();
+
+  if (!todos?.length) {
+    return null;
+  }
+
+  const activeTodosCount = todos.filter(todo => !todo.completed).length;
+  const completedTodosCount = todos.filter(todo => todo.completed).length;
```
Comment (Zibi95) [inline src/components/Footer/Footer.tsx:16]: I guess you don't have to filter out completedTodos. You can just now do some maths with activeTodosCount and all todos so 
`todos.length - activeTodosCount = completedTodosCount`

**Review by Zibi95 — APPROVED** [review]
Good job! However, I feel that you've overused custom hooks. While it might look clean because there is less code in the components, it can become really hard to understand when you need to debug the logic. Using custom hooks within other custom hooks can also lead to complex code. What I tend to do is create the logic without using a custom hook initially. Then, if I find a place where I need to reuse the logic, or if a significant amount of code is required to achieve the desired result, I create a custom hook. I can approve this task as you have met the requirements for this task, but try to rethink the logic and simplify the code for other task in this series. Don't skip react go as the roadmap goes. Redux is some extra and its useful if you build rly big apps.
